graph/patterns/alternate-key.md

```diff
@@ -0,0 +1,130 @@
+
+## When to Use this Pattern
+
+---
+
+This pattern works and makes sense when the alternate key is good enough to identify a single resource and provides an useful alternative to the client; while it does not work if the resultset has more than one element. In such case, the workload SHOULD return `400`
+
+## Example
+
+---
+
+The same user identified via the alternate key SSN, the canonical (primary) key ID using the non-canonical long form with specified key property name, and the canonical short form without key property name
+
+Declare `mail` and `ssn` as alternate keys on an entity:
+
+```xml
+<EntityType Name="User">
+   <Key>
+     <PropertyRef Name="id" />
+   </Key>
+   <Property Name="id" Type="Edm.Int32" />
+
+   <Property Name="mail" Type="Edm.String" />
+   <Property Name="ssn" Type="Edm.String" />
+   <Annotation Term="Keys.AlternateKeys">
+      <Collection>
+         <Record>
+            <PropertyValue Property="Key">
+               <Collection>
+                  <Record>
+                     <PropertyValue Property="Name" PropertyPath="mail" />
+                  </Record>
+               </Collection>
+            </PropertyValue>
+         </Record>
+         <Record>
+            <PropertyValue Property="Key">
+               <Collection>
+                  <Record>
+                     <PropertyValue Property="Name" PropertyPath="ssn" />
+                  </Record>
+               </Collection>
+            </PropertyValue>
+         </Record>
+      </Collection>
+   </Annotation>
+</EntityType>
+```
+
+1. Get a specific resource through `$filter`:
+
+```http
+GET https://graph.microsoft.com/v1.0/users/?$filter=(ssn eq '123-45-6789')
+```
+
+```json
+{
+  "value": [
+    {
+      "givenName": "Bob",
+      "jobTitle": "Retail Manager",
+      "mail": "bob@contoso.com",
+      "mobilePhone": "+1 425 555 0109",
+      "officeLocation": "18/2111",
+      "preferredLanguage": "en-US",
+      "surname": "Vance",
+      "userPrincipalName": "bob@contoso.com",
+      "id": "1a89ade6-9f59-4fea-a139-23f84e3aef66"
+    }
+  ]
+}
+```
+
+2. Get a specific resource through its primary key, and then through the two alternate keys:
+
+```http
+GET https://graph.microsoft.com/v1.0/users/1a89ade6-9f59-4fea-a139-23f84e3aef66
+GET https://graph.microsoft.com/v1.0/users(ssn='123-45-6789')
+GET https://graph.microsoft.com/v1.0/users(email='bob@contoso.com')
+```
+
+All of the 3 will yield the sare response:
+
+```json
+{
+  "givenName": "Bob",
+  "jobTitle": "Retail Manager",
+  "mail": "bob@contoso.com",
+  "mobilePhone": "+1 425 555 0109",
+  "officeLocation": "18/2111",
+  "preferredLanguage": "en-US",
+  "surname": "Vance",
+  "userPrincipalName": "bob@contoso.com",
+  "id": "1a89ade6-9f59-4fea-a139-23f84e3aef66"
+}
+```
+
+3. Requesting a resource through an alternate key which yields more than one result
```

Comment:
Per earlier comment - this should never happen.  The service should never make "name" as a valid alternate key.
Hence I think the 400 error is that "name" is not supported as an alternate key.

I would title this "Requesting a resource for an unsupported alternate key property"  or some such.

Reply:
Is this ensured by AGS or by the workload? In any case that is correct, I will reword this part

Reply:
That's a good question.  AGS has the schema and annotations, so *maybe* it could (and does) catch the use of a property that is not marked as an alternate key and save the round trip to the workload (fail fast).  But I don't know the answer.